\section{Overview}


This tutorial focuses on data formats used in \RevBayes.


\subsection*{Requirements}
We assume that you have read and hopefully completed the following tutorials:
\begin{itemize}
\item RB\_Getting\_Started
\end{itemize}



%
%
%
\newpage
\FloatBarrier
\section{Molecular Sequences}

\bigskip
\subsection{Getting Started}


\exs{Download data and output files from: \href{http://revbayes.github.io/tutorials.html}{http://revbayes.github.io/tutorials.html}}
\exs{Open the file \cl{primates\_cytb.nex} in your text editor. This file contains the sequences for the cytochrome B gene sampled from 13 species (Box 1). The elements of the \cl{DATA} block indicate the type of data, number of taxa, and length of the sequences.}


\begin{center}
Box 1: A fragment of the NEXUS file containing the ITS sequences for this exercise. \\
\end{center}
{\tt \scriptsize \begin{framed}
\begin{lstlisting}
#NEXUS 

Begin data;
Dimensions ntax=13 nchar=673;
Format datatype=DNA missing=? gap=-;
Matrix
Trig_excelsa   
TCGAAACCTG...
Fagus_engleriana   
TCGAAACCTG...
Fagus_crenata1   
TCGAAACCTG...
Fagus_japonica2   
TCGAAACCTG...
Fagus_japonica1   
TCGAAACCTG...
Fagus_orientalis   
TCGAAACCTG...
Fagus_sylvatica   
TCGAAACCTG...
Fagus_lucida1   
TCGAAACCTG...
Fagus_lucida2   
TCGAAACCTG...
Fagus_crenata2   
TCGAAACCTG...
Fagus_grandifolia   
TCGAAACCTG...
Fagus_mexicana   
TCGAAACCTG...
Fagus_longipetiolata   
TCGAAACCTG...
	;
End;
\end{lstlisting}
\end{framed}}


\subsection{Reading Molecular Sequences}


{\tt \begin{snugshade*}
\begin{lstlisting}
# show some general information
data
|*
|*
|*

# get some quick and useful attributes of the data
data.ntaxa()
data.names()	
\end{lstlisting}
\end{snugshade*}}



\subsection{Concatenating Sequences}


{\tt \begin{snugshade*}
\begin{lstlisting}
data_atpB <- readDiscreteCharacterData("data/conifer_atpB.nex")
data_atpB <- readDiscreteCharacterData("data/conifer_atpB.nex")
\end{lstlisting}
\end{snugshade*}}

Since the first step in this exercise is to assume a single model for both genes, we need to combine the two datasets.
Concatenate the two data matrices using the \cl{+} operator. This returns a single data matrix with both genes.

{\tt \begin{snugshade*}
\begin{lstlisting}
data <- concatenate(data_ITS, data_rbcL, data_matK)
\end{lstlisting}
\end{snugshade*}}



\subsection{Removing Sites or Genes}


\subsection{Removing Taxa}



\section{Morphological Data}



\section{Continuous Trait Data}


\section{Biogeographical Data}

\subsection{Nexus file}

The data file contains a matrix of binary characters corresponding to the observed ranges of the study taxa.

\noindent \\ \impmark  Open the file \texttt{examples/psychotria\_range.txt}.

\begin{framed}
\begin{lstlisting}
#NEXUS

begin data;
  dimensions ntax=19 nchar=4;
  format datatype=standard symbols = "01";
  matrix
    P_mariniana_Kokee2  1000
    P_mariniana_Oahu    0100
    ...
    P_hexandra_Oahu     0100
  ;
end;

Begin trees;
	TREE tree1 = ((((((((P_hawaiiensis_WaikamoiL1:0.9656850499,
	P_mauiensis_Eke:0.9656850499):0.7086257935,(P_fauriei2:1.23
	0218511,P_hathewayi_1:1.230218511):0.4440923324):0.17671155
	...
	89):0.4630447802,P_hexandra_Oahu:2.826939991):2.372081244);
End;
\end{lstlisting}
\end{framed}

Range data is stored in standard Nexus format.
In the {\tt data} block, the first line gives the dimensions of the data matrix and the second line indicates we will be using binary characters.
The four characters correspond to areas defined by the geography file (next subsection).
Rows in the {\tt matrix} block correspond to taxa and their range data, while columns give in which areas each taxon is present (1) or absent (0).
For example, taxon {\tt P\_hexandra\_Oahu} is present only in area 2.

The {\tt trees} block gives the tree describing the shared ancestry of the study species.
Because range evolution occurs in units of geological time, the analysis in this tutorial requires a high-quality time-calibrated phylogeny.
This typically requires a multiple sequence alignment over several loci plus fossils for calibration.
Since this data availability is often the limiting factor for which taxa to include for your analysis, it is best to produce the phylogeny first.
Only afterwards should you begin assembling data for your data matrix.
If your phylogeny cannot be calibrated (e.g. it has no fossils) your best alternative is to proceed with a time tree resulting from a divergence time estimation analysis.
For this tutorial, the phylogeny is assumed to contain no uncertainty.

\subsection{Atlas file}

The geography used in this tutorial represents the Hawaiian archipelago.
Beneath Hawaii, currently the largest and youngest island, is a volcanic hotspot that periodically creates new islands.
The ages of these islands are fairly well known, meaning we can model range availability as a function of time.
Following \citet{ree08}, we will lump groups of smaller islands into single areas to simplify the analysis, leaving us with four areas: Hawaii (H), Oahu (O), Maui (M; this includes Molokai and Lanai), and Kauai (K; this includes Niihau).
These areas are modeled have arisen 0.5, 1.9, 3.7, and 5.5 million years ago, respectively.

Although the model will use discrete-state biogeographic ranges, geographical area is naturally continuous.
This means we must impose some discretization upon the geography to designate a set of biogeographically meaningful characters called areas.
Different methods use different criteria for this discretization, so it is best to perform the discretization yourself rather than blindly using the discretization given from a previous study or method (but do blindly use the dataset included in this tutorial).
Some geographies have natural discretizations: for instance, the Hawaiian archipelago forms naturally discrete areas on the basis of islands.
For many geographies, however, it may unclear how to perform this discretization.
Much like morphological analyses, you might choose to choose areas based on expert opinion, based on some model, or using some ``naive'' uniform discretization.
This procedure is not part of the tutorial, but you should be aware that area definitions are not always obvious or objective.


\newpage

\noindent \\ \impmark  Open the file \texttt{examples/hawaii\_dynamic.atlas.txt}.

%\begin{framed}
%\begin{lstlisting}[]%[basicstyle=\tiny \listingsfont, columns=texcl]
%{
%	"name":"HawaiianArchipelago5my",
%	"epochs": [
%		{
%			"name":"epoch1",
%			"start_age":100.0,
%			"end_age":3.7,
%			"areas":
%			[{ "name":"Kauai",  "latitude":19.5667, "longitude": -155.5000, "dispersalValues": [ 1,0,0,0 ] },
%			{ "name": "Oahu",   "latitude":19.5667, "longitude": -155.5000, "dispersalValues": [ 0,0,0,0 ] },
%			{ "name": "Maui",   "latitude":19.5667, "longitude": -155.5000, "dispersalValues": [ 0,0,0,0 ] },
%			{ "name": "Hawaii", "latitude":19.5667, "longitude": -155.5000, "dispersalValues": [ 0,0,0,0 ] }]
%		},
%		{
%			"name":"epoch2",
%			...
%		},
%		{
%			"name":"epoch3",
%			...
%		},
%		{
%			"name":"epoch4",
%			"start_age":0.5,
%			"end_age":0.0,
%			"areas":
%			[{ "name":"Kauai",  "latitude":22.0833, "longitude": -159.5000, "dispersalValues": [ 1,1,1,1 ] },
%			{ "name": "Oahu",   "latitude":21.4722, "longitude": -157.9772, "dispersalValues": [ 1,1,1,1 ] },
%			{ "name": "Maui",   "latitude":20.8000, "longitude": -156.3333, "dispersalValues": [ 1,1,1,1 ] },
%			{ "name": "Hawaii", "latitude":19.5667, "longitude": -155.5000, "dispersalValues": [ 1,1,1,1 ] }]
%		}
%	]
%}
%\end{lstlisting}
%\end{framed}

This is called the Atlas file, which uses a file format called JSON.
JSON is a lightweight format used to assign values to variables in a hierarchical manner.
There are three main tiers to the hierarchy in the Atlas file: the atlas, the epoch, and the area.
In the lowest tier, each area corresponds to a character in the model and is assigned it's own properties.
In the middle tier, each epoch contains the set of homologous areas (characters) that may be part of a species' range, but importantly the properties of these areas may take on different values during different intervals of time, as given by the {\tt start\_age} and {\tt end\_age} variables.
Because the tree and range evolution model also operate on units of geological time, the rates of area gain and loss can condition on areas' properties as a function of time.
Sometimes these models are called stratified models or epochal models.
Finally, the atlas contains the array of epochs in the highest tier.

Each area is assigned a {\tt latitude} and {\tt longitude} to represent its geographical coordinates, ideally the area's centroid.
If a centroid does not represent the distance between areas, splitting the area into multiple smaller areas is reasonable.
The data augmentation approach used in this analysis allows you to use more areas as desired, whereas matrix exponentiation methods are limited to approximately ten areas.
The distance between any two areas' coordinates inform to distance-dependent dispersal parameter ($\beta$ from the $\eta(\cdot)$ function) for range expansion events, so coordinates roughly close to the center of the area suffice.
Here, the {\tt latitude} and {\tt longitude} change in each of the four epochs, where they begin at the current location of Hawaii and drift northwesterly until they reach their current positions.

In addition, each area is marked as habitable or not using the {\tt dispersalValues} array.
The elements in the array correspond to the other areas defined in the analysis.
For example, in {\tt epoch1}, Kauai's {\tt dispersalValues} is equal to {\tt [ 1,0,0,0 ]}, which indicates Kauai exists at that point in time but it is not in contact with any other areas, i.e. the range in that area cannot expand into other areas.
The {\tt dispersalValues} for Oahu, Maui, and Hawaii are all equal to {\tt [ 0,0,0,0 ]}, meaning no species may be present in that area during the time interval of {\t epoch1} during ages from 10.0 to 3.7. In contrast, {\tt epoch4}, from ages 0.5 to the present, range expansions may occur between any pair of areas and any area may be included in a species' range.


\vspace{5cm}
Questions about this tutorial can be directed to: \\\vspace{-10mm}
\begin{itemize}
\item Tracy Heath (email: \href{mailto:tracyh@berkeley.edu}{tracyh@berkeley.edu}) \\\vspace{-8mm}
\item Michael Landis (email: \href{mailto:mlandis@berkeley.edu}{mlandis@berkeley.edu}) \\\vspace{-8mm} 
\item Sebastian H\"{o}hna (email: \href{mailto:sebastian.hoehna@gmail.com}{sebastian.hoehna@gmail.com}) \\\vspace{-8mm}
\item Brian R. Moore (email: \href{mailto:brianmoore@ucdavis.edu}{brianmoore@ucdavis.edu}) \\\vspace{-8mm}
\end{itemize}


\bibliographystyle{mbe}
\bibliography{bib_tex/master_refs}
