\section{Overview of Chromosome Number Evolution Models} \label{sec:chromo_basic_intro}


Chromosome changes
represent major evolutionary mechanisms that have long been a focal point of study.
Changes in chromosome number such as the gain or loss of a single
chromosome (dysploidy), or the doubling of the entire genome (polyploidy),
can have phenotypic consequences,
affect the rates of recombination,
and increase reproductive isolation among lineages and thus drive diversification \citep{stebbins1971chromosomal}.
Recently, evolutionary biologists have increasingly studied the macroevolutionary
consequences of chromosome changes within a molecular phylogenetic framework,
mostly utilizing the likelihood-based ChromEvol models of chromosome number evolution
introduced by \citet{mayrose2010probabilistic}.
The ChromEvol models have permitted phylogenetic studies
of ancient whole genome duplication events,
rapid ``catastrophic'' chromosome speciation,
major reevaluations of the evolution of angiosperms,
and new insights into the fate of polyploid lineages \citep[e.g.][]{pires2008renaissance, mayrose2011recently, tank2015nested}.
The basic ChromEvol model has been extended
to examine the association of phenotype with chromosome evolution \citep[BiChroM;][]{zenil2017testing},
and to incorporate cladogenetic changes and diversification rates \citep[ChromoSSE;][]{freyman2016cladogenetic}.

Here we describe the ChromEvol model as implemented in \RevBayes,
which except for one detail noted below 
is the same mathematical model introduced in \citet{mayrose2010probabilistic}.
In further sections, we will show how to set up extensions
such as BiChroM and ChromoSSE which build on the useful but basic ChromEvol model of chromosome number evolution.


\subsection{The ChromEvol Model}


In ChromEvol the evolution of chromosome number
is represented as a continuous-time Markov process,
similar to models of molecular evolution and discrete morphological evolution.
The continuous-time Markov process is described by
an instantaneous rate matrix $Q$ where the value of each element
represents the instantaneous rate of change within a lineage
from a genome of $i$ chromosomes to a genome of $j$ chromosomes.
For all elements of $Q$ in which either $i = 0$ or $j = 0$
we define $Q_{ij} = 0$.
For the off-diagonal elements $i \neq j$ with positive values of $i$ and $j$,
$Q$ is determined by:
\begin{equation} \label{eq:anagenetic1}
Q_{ij} = 
    \begin{cases}
        \gamma_a                          & j = i + 1,    \\
        \delta_a                          & j = i - 1,    \\
        \rho_a                                & j = 2i,       \\
        \eta_a                                 & j = 1.5i,     \\
        0                                   & \mbox{otherwise},   
    \end{cases}
\end{equation}
where $\gamma_a$, $\delta_a$, $\rho_a$, and $\eta_a$ are the
rates of chromosome gains, losses, polyploidizations, and
demi-polyploidizations.
We use the subscript $a$ for all these rates to differentiate the rates between anagenetic ($a$) and cladogenetic ($c$) events (see the ChromoSSE model in a later section).

If we are interested in modeling scenarios in which
the probability of fusion or fission events
are positively or negatively correlated with the number of chromosomes
we can define $Q$ as:
\begin{equation} \label{eq:anagenetic2}
Q_{ij} = 
    \begin{cases}
        \gamma_a e^{\gamma_m (i - 1)}  & j = i + 1,    \\
        \delta_a e^{\delta_m (i - 1)}     & j = i - 1,    \\
        \rho_a                                & j = 2i,       \\
        \eta_a                                 & j = 1.5i,     \\
        0                                   & \mbox{otherwise},   
    \end{cases}
\end{equation}
where $\gamma_m$ and $\delta_m$ are rate modifiers
of chromosome gain and loss, respectively,
that allow the rates of chromosome gain and loss to depend
on the current number of chromosomes.
If the rate modifier $\gamma_m = 0$, then
$\gamma_a e^{0 (i - 1)} = \gamma_a$.
If the rate modifier $\gamma_m > 0$, then
$\gamma_a e^{\gamma_m (i - 1)} \geq \gamma_a$ (\IE rates increase with more chromosomes),
and if $\gamma_m < 0$ then
$\gamma_a e^{\gamma_m (i - 1)} \leq \gamma_a$ (\IE rates decrease with more chromosomes).
Note that this parameterization differs slightly from the original
ChromEvol model; here we assume the rates of chromosome change
can vary exponentially as a function of the current chromosome number,
whereas ChromEvol as originally described by \citet{mayrose2010probabilistic} assumes
a linear function.
The theoretical reasons for this difference are described in \citet{freyman2016cladogenetic},
however in practice on most empirical datasets the difference appears insignificant.


Demi-polyploidization is the union of
a reduced and an unreduced gametes that produces a cytotype 
with 1.5 times the number of chromosomes.
The number of chromosomes in a genome must of course be an integer, 
so for odd values of $i$, $Q_{ij}$ is set to $\eta/2$
for the two integer values of $j$ resulting when $j = 1.5i$ is rounded up and down.


As in all continuous-time Markov models,
the diagonal elements $i = j$ of $Q$ are defined as:
\begin{equation}
    Q_{ii} = - \sum_{i \neq j}^{C_m} Q_{ij}.
\end{equation}
The probability
of anagenetically evolving from chromosome number $i$ to $j$ along a branch
of length $t$ is then calculated by
exponentiation of the instantaneous rate matrix:
\begin{equation} \label{eq:anagenetic_probs}
    P_{ij}(t) = e^{-Qt}.
\end{equation}
Given a phylogeny and chromosome counts of the extant lineages,
this model can be used in either a maximum likelihood or Bayesian inference
framework to estimate the rates of chromosome change and the ancestral
chromosome numbers.

\subsection{Hypothesis Testing and Model Uncertainty}

The ChromEvol model described above is actually a class of models;
for example we could exclude demi-polyploidization by fixing it's rate to 0.
A common use for different models of chromosome evolution is to test hypotheses.
For example, are polyploidization events
occuring primarily at speciation events and possibly driving diversification,
or do polyploidization events occur within lineages and unassociated with lineage splitting?
To answer this, one could use \RevBayes to set up two different models, one allowing cladogenetic
polyploidization (see Section \ref{subsect:clado_simple}) and a second
using a model with only anagenetic polyploidization (like the ChromEvol model described above).
One could then calculate a Bayes factor to compare which model better explained the observed data.
See the \RevBayes tutorial \textbf{Compute Marginal Likelihoods to Select Between Models}
for more information on how to calculate Bayes factors in \RevBayes.


Another option in \RevBayes is to use Bayesian model averaging.
Out of the class of all chromosome evolution models
described here, it is possible that no single model
will adequately describe the chromosome evolution
of a given clade.
To explore the entire space of all possible models of chromosome number evolution 
one could specify a reversible jump Markov chain Monte Carlo \citep{green1995reversible} that samples across models of different dimensionality,
drawing samples
from chromosome evolution models in proportion to their posterior probability and
enabling Bayes factors for each model to be calculated.
This approach incorporates model uncertainty
by permitting model-averaged inferences that do not condition on a single model;
we draw estimates of ancestral chromosome numbers and rates of chromosome evolution
from all possible models weighted by their posterior probability.
For general reviews of this approach to model averaging see
\cite{madigan1994model}, \cite{hoeting1999bayesian}, \cite{kass1995bayes},
and for its use in phylogenetics see \citet{posada2004model}.
Averaging over all models has been shown to
provide a better average predictive ability
than conditioning on a single model
\citep{madigan1994model}.
Conditioning on a single model ignores model uncertainty,
which can lead to an underestimation in the uncertainty
of inferences made from that model \citep{hoeting1999bayesian}.
In our case, this can lead to overconfidence in estimates
of ancestral chromosome numbers and chromosome
evolution parameter value estimates.
For details on how to implement Bayesian model averaging in \RevBayes
with chromsome evolution see \citet{freyman2016cladogenetic}.



\subsection{Next Steps}


The basic ChromEvol model as described above can be extended in a number of useful ways 
that will be covered in further sections.
In the next section, however, we'll set up and run a simple \RevBayes analysis using the ChromEvol model
before moving on to the more complex models.


\newpage
