\section{Dispersal-Extinction-Cladogenesis model}

\subsection{Model and method}

First, we define the range for taxon $i$ as the bit vector $X_i$, where $X_{i,j} = 1$ if the taxon is present in area $j$ and $X_{i,j} = 0$ if the taxon is absent.
Each taxon range is a bit vector of length $N$ areas.
For example, if taxon $B$ is present only in areas 2 and 3 out of $N=3$ areas, its range is represented as $X_B = (0,1,1)$, which is translated to the bit string $X_B=011$ for short.
If we apply the labels A, B, and C to our three areas, this bit vector can also be represented as a set: $X_B=011$ is the same as $X_B = \{ B, C \}$, or $X_B=BC$ for short.
The data matrix, $\textbf{X}$, is analogous to a multiple sequence alignment where each element in the data matrix reports a discrete value for a homologous character shared by all taxa at column $j$.

\subsection{Modeling anagenic range evolution}

Next, we need a model of anagenic range evolution.
Since we have discrete characters we'll use the continuous-time Markov chain, which allows us to compute transition probability of a character changing from $i$ to $j$ in time $t$ through matrix exponentiation
\[
\mathbf{P}_{i,j}(t) = \left[ \exp \left\lbrace \mathbf{Q}t \right\rbrace \right]_{i,j},
\]
where $\textbf{Q}$ is the instantaneous rate matrix defining the rates of change between all pairs of characters, and $\textbf{P}$ is the transition probability rate matrix.
This technique of matrix exponentiation is powerful because it integrates over all possible scenarios of character transitions that could occur during $t$ so long as the chain begins in state $i$ and ends in state $j$. Remember, $i$ and $j$ represent different ranges, each of which is encoded as a set of occupied areas.

We can then encode range evolution events into the allowed character transitions of $\textbf{Q}$ and parameterize the events so that we may infer their relative importance to generating our observed ranges.
We'll take a simple model of range expansion (e.g. $BC \rightarrow ABC$) and range contraction (e.g. $BC \rightarrow C$).
(Range expansion may also be referred to as dispersal or area gain and range contraction as extirpation or area loss.)
The rates in the transition matrix for three areas might appear as

\[
\textbf{Q} = 
	\begin{array}{c|cccccccc}
		& \emptyset & A & B & C & AB & AC & BC & ABC \\
		\hline
		\emptyset & - & 0 & 0 & 0 & 0 & 0 & 0 & 0 \\
		A & e_A & - & 0 & 0 & d_{AB} & d_{AC} & 0 & 0 \\
		B & e_B & 0 & - & 0 & d_{BA} & 0 & d_{BC} & 0 \\
		C & e_C & 0 & 0 & - & 0 & d_{CA} & d_{CB} & 0 \\
		AB & 0 & e_B & e_A & 0 & - & 0 & 0 & d_{AC} + d_{BC} \\
		AC & 0 & e_C & 0 & e_A & 0 & - & 0 & d_{AB} + d_{CB} \\
		BC & 0 & 0 & e_C & e_B & 0 & 0 & - & d_{BA} + d_{CA} \\
		ABC & 0 & 0 & 0 & 0 & e_C & e_B & e_A & - \\								
	\end{array},
\]
where $e = ( e_A, e_B, e_C )$ are the (local) extinction rates per area, and $d = ( d_{AB}, d_{AC}, d_{BC}, d_{CB}, d_{CA}, d_{BA})$ are the dispersal rates between areas.

{\bf Q: For the three-area DEC rate matrix above, what is the rate of leaving state 101? That is, what is the absolute value of the diagonal term in the rate matrix for $Q_{AB,AB}$? }

Note the rate of more than one event occurring simultaneously is zero, so a range must expand twice by one area in order to expand by two areas.

{\bf Q: What events might explain a transition from range $ABC$ to range $A$? From range $AB$ to range $C$?}

Of course, this model can be specified for more than three areas.

{ \bf Q: Imagine a DEC rate matrix with four areas, $ABCD$. What would be the dispersal rate for $Q_{BC,BCD}$? How many states does a DEC rate matrix with four areas have? What is the relationship between the number of areas and the number of states under the DEC model? }

Let's consider what happens to the size of \textbf{Q} when the number of areas, $N$, becomes large.
For three areas, \textbf{Q} is size $8 \times 8$.
For ten areas, \textbf{Q} is size $2^{10} \times 2^{10} = 1024 \times 1024$, which approaches the largest size matrices that can be exponentiated in a practical amount of time.
For biogeographic inference under large numbers of areas, see the Tutorial XXX \citep{landis13}.


\subsection{Modeling cladogenic range evolution}

In addition to dispersal and extinction, the DEC models cladogenic range evolution events.
For each internal node in the reconstructed tree, one of two cladogenic events can occur: sympatry or allopatry.
Say the range of a species approaching an internal node, i.e. that is about to speciate, is $A$.
Since the species range is size one, this always results in sympatry, where both daughter lineages inherit the ancestral species range, so both lineages begin in state $A$.
The notation $A \rightarrow A \mid A$ describes this event: the state is $A$ before cladogenesis, the left daughter inherits range $A$ after cladogenesis, as does the right daughter.

Now suppose the ancestral range is $ABC$.
Under sympatric cladogenesis, one lineage identically inherits the ancestral species range, $ABC$, while the other lineage inherits only a single area, i.e. only $A$ or $B$ or $C$.

Under allopatric cladogenesis, the ancestral range is split evenly among daughter lineages, e.g. one lineage may inherit $AB$ and the other inherits $C$.
Both daughter lineages inherit the entire ancestral species range following widespread sympatric cladogenic events.
For a general description of state transitions for cladogenic events, see \citet{matzke13}.

{\bf Q: What are the 6 possible states in the daughter lineages after cladogenesis given the state is $AB$ before cladogenesis?}

The probabilities of anagenic change along lineages must account for all combinations of starting states and ending states.
For 3 areas, there are 8 states, and thus $8 \times 8 = 64$ probability terms for pairs of states.
For cladogenic change, we need transition probabilities for all combinations of states before cladogenesis, after cladogenesis for the left lineage, and after cladogenesis for the right lineage.
Like above, for three areas, there are 8 states, and $8 \times 8 \times 8 = 512$ cladogenic probability terms.

{\bf Q: For three areas, there are 3 + 12 + 6 = 21 possible sympatry events and 6 + 6 = 12 possible allopatry events. How many terms in the cladogenesis matrix are zero?}

The DEC model ignores speciation events hidden by extinction or incomplete taxon sampling.
The probability of cladogenesis and local extinction events would ideally be linked to a birth-death process, as it is in the GeoSSE model \citep{goldberg11}.
Unfortunately, since the numerical method for SSE models scale poorly, and DEC models remain the only option when the geography has more than two or three areas.
For more than ten areas, we need data augmentation to integrate over range evolution, which is used in Section XXX.

\begin{figure}[H]
\centering
\includegraphics[width=5in]{figures/bg_dec_dag}
\caption{Graphical model of DEC. The tree plate's topology is fixed by $T$, where each internal node has both an anagenic and cladogenic random variable ($X_i^{\text{an}}$ and $X_i^{\text{cl}}$, resp.) that represents an ancestral species before and after it speciated. Anagenic change is modeled by a continuous time Markov process, where $Q_{DE}$ is the instantaneous rate matrix of area gain and loss, as parameterized by $\lambda$. The geographic distance rate modifier function, $\eta$, takes in the geographical distances and strata as $G$, and the distance power parameter, $\beta$. Cladogenic change is modeled by $P_C$, a Dirichlet-distributed simplex with a flat prior.}
\end{figure}

The rest of this section will describe how to run a simple DEC analysis using RevBayes.

\newpage

%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%

\subsubsection{Specifying a simple DEC model}

We'll use a dataset for 23 primates, which are the same taxa as used in Section XXX. To keep the model simple, we'll ranges are over three areas: the New World, Africa, and Eurasia.
For simplicity, we'll assume their phylogeny is time-calibrated, errorless, and fixed.

First, set your working directory.

\begin{snugshade}
\begin{lstlisting}
setwd("/Users/arwallace/projects/RB_Biogeography_tutorial/")
\end{lstlisting}
\end{snugshade}

First we'll create some {\tt String} variables for file handling

\begin{snugshade}
\begin{lstlisting}
fp = "./"
data_fn = fp + "data/primates_bg_n3.tsv"
tree_fn = fp + "data/primates.tree"
\end{lstlisting}
\end{snugshade}

then read in our character data

\begin{snugshade}
\begin{lstlisting}
data = readTSVCharacterData(data_fn, type="NaturalNumbers")
\end{lstlisting}
\end{snugshade}

and our tree

\begin{snugshade}
\begin{lstlisting}
tree <- readTrees(tree_fn)[1]
\end{lstlisting}
\end{snugshade}

Next, compute the number of states from the number of areas

\begin{snugshade}
\begin{lstlisting}
n_areas = data.nchar()
n_states = 2^data.nchar()
\end{lstlisting}
\end{snugshade}

Declare index variables for our move and monitor vectors for future use

\begin{snugshade}
\begin{lstlisting}
mvi = 1
mni = 1
\end{lstlisting}
\end{snugshade}

Now, we'll begin to construct the rate matrix for anagenic events. First create a matrix, 8-by-8 in size, initialized with all zeroes

\begin{snugshade}
\begin{lstlisting}
for (i in 1:n_states) {
    for (j in 1:n_states) {
        r[i][j] <- 0.
    }
}
\end{lstlisting}
\end{snugshade}

Now we need to populate the non-zero rate matrix elements, which are in terms of dispersal and extinction rates. We'll use one dispersal rate and one extinction rate for this tutorial, and explore more complex models in later sections.

First, create a extinction rate parameter and assign it a scale move

\begin{snugshade}
\begin{lstlisting}
r_e ~ dnExp(10.)
mv[mvi++] = mvScale(r_e, weight=5)
\end{lstlisting}
\end{snugshade}

Before assigning the rates to the rate matrix, we'll create a vector to hold the per-area extinction rates 

\begin{snugshade}
\begin{lstlisting}
for (i in 1:n_areas) {
    e[i] := r_e
}
\end{lstlisting}
\end{snugshade}

Now create the dispersal rate and scale move

\begin{snugshade}
\begin{lstlisting}
r_d ~ dnExp(10.)
mv[mvi++] = mvScale(r_d, weight=5)
\end{lstlisting}
\end{snugshade}

then assign the between-area dispersal rates as determined by {\tt $r_d$}

\begin{snugshade}
\begin{lstlisting}
for (i in 1:n_areas) {
    for (j in 1:n_areas) {
        if (i != j) {
            d[i][j] := r_d
        }
    }
}
\end{lstlisting}
\end{snugshade}

Next, we'll populate the non-zero rate matrix elements.
Rates are indexed by the natural number value of the range, e.g. the range spanning Eurasia and Africa is coded as 011, which is state 4.

First assign the extinction (range loss) rates

\begin{snugshade}
\begin{lstlisting}
r[4][2] := e[2]                 # 011 -> 001 : Extirpate in area 2
r[4][3] := e[3]                 # 011 -> 010 : Extirpate in area 3
r[6][2] := e[1]                 # 101 -> 001 : Extirpate in area 1
r[6][5] := e[3]                 # 101 -> 100 : Extirpate in area 3
r[7][3] := e[1]                 # 110 -> 010 : Extirpate in area 1
r[7][5] := e[2]                 # 110 -> 100 : Extirpate in area 2
r[8][4] := e[1]                 # 111 -> 011 : Extirpate in area 1
r[8][6] := e[2]                 # 111 -> 101 : Extirpate in area 2
r[8][7] := e[3]                 # 111 -> 110 : Extirpate in area 3
\end{lstlisting}
\end{snugshade}

then the dispersal (range gain) rates

\begin{snugshade}
\begin{lstlisting}
r[2][4] := d[3][2]              # 001 -> 011 : Disperse from area 3 to 2
r[2][6] := d[3][1]              # 001 -> 101 : Disperse from area 3 to 1
r[3][4] := d[2][3]              # 010 -> 011 : Disperse from area 2 to 3
r[3][7] := d[2][1]              # 010 -> 110 : Disperse from area 2 to 1
r[5][6] := d[1][3]              # 100 -> 101 : Disperse from area 1 to 3
r[5][7] := d[1][2]              # 100 -> 110 : Disperse from area 1 to 2
r[4][8] := d[2][1] + d[3][1]  # 011 -> 111 : Disperse from area 2 to 1 and from 3 to 1
r[6][8] := d[1][2] + d[3][2]  # 101 -> 111 : Disperse from area 1 to 2 and from 3 to 2
r[7][8] := d[1][3] + d[2][3]  # 110 -> 111 : Disperse from area 1 to 3 and from 2 to 3
\end{lstlisting}
\end{snugshade}

Show the value of {\tt r} and compare it to the matrix in Equation (XX).

Of course we did not need to declare {\tt d} and {\tt e} to assign {\tt r}, but we'll see these intermediate variables act as a template expose the structure of {\tt r} for modification.

So far, we only have the desired parameterization of the rate matrix, but we still haven't created a rate matrix function. Converting the vector-of-vectors, {\tt r}, into a simplex allows us to use existing rate matrix functions. 

First, we'll convert {\tt r} into a one-dimensional vector, skipping the diagonal elements.

\begin{snugshade}
\begin{lstlisting}
k <- 1
for (i in 1:n_states) {
    for (j in 1:n_states) {
        if (i != j) {
            er_nat[k] := r[i][j]
            k += 1
        }
    }
}
\end{lstlisting}
\end{snugshade}

Finally, normalize {\tt er\_nat} using a simplex, then pass the resulting exchangeability rates as arguments into the rate matrix function, {\tt q}.

\begin{snugshade}
\begin{lstlisting}
er := simplex(er_nat)
bf <- simplex(rep(1,n_states))
q := fnFreeK(er, bf)
\end{lstlisting}
\end{snugshade}

This yields the desired three-area DEC rate matrix modeling anagenic character change. In contrast, cladogenic event probabilites are given by a transition probability matrix and do not require a rate matrix.

First, we will create a vector of prior weights on cladogenesis events. Here, we assign a flat prior to all cladogenic events

\begin{snugshade}
\begin{lstlisting}
widespread_sympatry_wt <- 1.0
subset_sympatry_wt     <- 1.0
allopatry_wt           <- 1.0
clado_prior            <- [ widespread_sympatry_wt, subset_sympatry_wt, allopatry_wt ]
\end{lstlisting}
\end{snugshade}

then create the distribution over cladogenic event types and add its MCMC move

\begin{snugshade}
\begin{lstlisting}
clado_type             ~ dnDirichlet(clado_prior)
mv[mvi++]               = mvSimplexElementScale(clado_type, alpha=10, weight=5)
\end{lstlisting}
\end{snugshade}

To give the simplex elements descriptive names when monitored, assign the values to deterministic nodes

\begin{snugshade}
\begin{lstlisting}
widespread_sympatry    := clado_type[1]
subset_sympatry        := clado_type[2]
allopatry              := clado_type[3]
\end{lstlisting}
\end{snugshade}

Then create the cladogenic transition probability matrix

\begin{snugshade}
\begin{lstlisting}
b <- simplex(1,1)
clado_prob := fnCladoProbs(clado_type_prob, b, n_areas, 2)
\end{lstlisting}
\end{snugshade}

Add a parameter for a biogeographical clock, which scales the overall rate of range evolution.
As a prior, an exponential distribution with rate 10 generates one dispersal or extinction event per 10 million years.

\begin{snugshade}
\begin{lstlisting}
clock_bg ~ dnExp(10)
mv[mvi++] = mvScale(clock_bg, weight=5)
\end{lstlisting}
\end{snugshade}

Finally, all our model components are encapsulated in the {\tt dnPhyloCTMCClado} distribution, which is similar to {\tt dnPhyloCTMC} except specialized to integrate over cladogenic events. Although this dataset has three areas, it is recognized single character with states valued from 1 to $2^3$, hence {\tt nSites=1}.

\begin{snugshade}
\begin{lstlisting}
m ~ dnPhyloCTMCClado( tree=tree, Q=q, cladoProbs=clado_prob, branchRates=clock_bg, nSites=1, type="NaturalNumbers" )
\end{lstlisting}
\end{snugshade}

The remaining tasks should be familiar by now, so we can proceed briskly. Attach the observed ranges to the model.

\begin{snugshade}
\begin{lstlisting}
m.clamp(data)
\end{lstlisting}
\end{snugshade}

Compose the model.

\begin{snugshade}
\begin{lstlisting}
mdl = model(m)
\end{lstlisting}
\end{snugshade}

Add the monitors.

\begin{snugshade}
\begin{lstlisting}
mn[mni++] = mnScreen(clock_bg, d[1][2], d[1][3], d[2][1], d[2][3], d[3][1], d[3][2], e[1], e[2], e[3], widespread_sympatry, subset_sympatry, allopatry)
mn[mni++] = mnFile(clock_bg, d[1][2], d[1][3], d[2][1], d[2][3], d[3][1], d[3][2], e[1], e[2], e[3], widespread_sympatry, subset_sympatry, allopatry, file=fp + "output/out.txt")
\end{lstlisting}
\end{snugshade}

Create the MCMC object, and run the chain after burn-in.
\begin{snugshade}
\begin{lstlisting}
ch = mcmc(mv,mn,mdl)
ch.burnin(1000, 10)
ch.run(10000)
\end{lstlisting}
\end{snugshade}

\subsubsection{Per-area rates}

Biologically, local extinction events probably do not occur at equal rates across all areas, as done above.
Ecological factors, geographical distances, etc. might cause these parameters to be weakly correlated or completely uncorrelated.
Dispersal rates, also, might not be the same between pairs of areas, or even symmetric depending on the direction of dispersal.
Rather than constraining all events of a type to share a common rate, instead you might give each area it's own extinction parameter

\begin{snugshade}
\begin{lstlisting}
for (i in 1:3) {
    e[i] ~ dnExp(10.)
    mv[mvi++] = mvScale(e, weight=5)
}
\end{lstlisting}
\end{snugshade}

or give each ordered pair of areas a it's own dispersal rate

\begin{snugshade}
\begin{lstlisting}
for (i in 1:3) {
    for (j in 1:3) {
        if (i != j) {
            d[i][j] ~ dnExp(10.)
            mv[mvi++] = mvScale(d[i][j], weight=5)
        }
    }
}
\end{lstlisting}
\end{snugshade}

This parameterization is identical to matrix introduced in Eq XXX.
In Section XXX, we'll extend this idea further to parameterize features of range evolution, and incorporate paleogeological information.

\subsubsection{Exercises}

\begin{itemize}
\item Widespread sympatric speciation is thought to be evolutionarily rare. Set the Dirichlet prior on cladogenic event types to heavily disfavor these events. Using Tracer, describe how changing the cladogenesis prior affects the extinction rate when compared with the ``common rate'' model.
\item (1 of 2) Saving your commands to a file, create a script to produce the ``per-area'' rate model.
\item (2 of 2) Determine if the data support the ``common rate'' model over the ``per-area'' rate model. Use the stepping stone method to compute marginal likelihoods, which will let you compute Bayes factors for model selection (see Section XXX).
\item (Advanced) Using what you learned in this tutorial and the CTMC tutorial (Section XXX), perform a joint analysis of molecular and biogeographic evolution for primates. Introduce a common prior for the molecular and biogeographical clocks, e.g.
\begin{snugshade}
\begin{lstlisting}
clock_mol       ~ dnExp(10.)
clock_scale_bg  ~ dnGamma(2.,2.)
clock_bg        := clock_mol * clock_scale_bg
\end{lstlisting}
\end{snugshade}
\end{itemize}



\bibliography{bayes}